### ISSUE TITLE
CakePHP 3 - List finder on association using contain

### ISSUE THREAD

**Issue by cnizzardini** [cakephp/cakephp#11644]
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: v3.4.13
* Platform and Target: Apache 2.x PHP 7.x

### What you did
        $query = $this->find()
            ->contain([
                'PropertyDiscountOption',
                'PropertyDiscount' => function ($q) {
                    return $q->find('list',[
                        'keyField' => 'property_id',
                        'valueField' => 'discount_id'
                    ]);
                }
            ]);

### What happened
While the associated query for Property Discount does appear to run when I debug it, the list is never associated back to its parent association. For some reason list fails within a containable.

### What you expected to happen
I expect to be able to have a list of PropertyDiscount associated to the main query. Is Cake having difficulty associating this back to its parent association when I specify a list finder?

I have indeed Googled for solutions and posted this problem in cakesf.slack.com, which leads me to believe this may be past a simple question.

**Comment by dereuromark** [comment]
Are you using non standard singular names on purpose? For those related model names.

**Comment by cnizzardini** [comment]
It's legacy. I am doing this because I migrated from Code Igniter 2 to CakePHP 3. The original engineer was a fan of singular names. Obviously, its not worth migrating so many tables, so I made concessions.

**Comment by dereuromark** [comment]
OK, and did you check if the same issue still exists on latest 3.5.x? Since often times these things could already be fixed on a newer minor or patch version (which should always been upgraded to rather sooner than later).

**Comment by cnizzardini** [comment]
I bumped to v3.5.10, does not appear to be working.

**Comment by cnizzardini** [comment]
The workaround, though not great, appears to be using the following in the ORM in place of the list finder:

`return $q->select(['property_id','discount_id'])->hydrate(false)->find('all');`

And then later on using Hash extract: 

`$properties = Hash::extract($discount->property_discount, '{n}.property_id');`

It would be best to just have a key-value pair from the ORM so I can use more efficient native functions like isset() or array_key_exists(). That's my use-case at least.

**Comment by markstory** [comment]
Does your `find('list')` operation fetch the foreign keys require to set relational properties? Without foreign keys the ORM cannot stitch your data into entities.

**Comment by cnizzardini** [comment]
@markstory I do believe I have done this correctly. Here are the associations from PropertyDiscountTable

    public function initialize(array $config)
    {
        parent::initialize($config);

        $this->table('property_discount');
        $this->displayField('id');
        $this->primaryKey('id');

        $this->addBehavior('Timestamp');

        $this->belongsTo('Property', [
            'foreignKey' => 'property_id',
            'joinType' => 'INNER'
        ]);
        $this->belongsTo('Discount', [
            'foreignKey' => 'discount_id',
            'joinType' => 'INNER'
        ]);
    }

And from DiscountTable

        $this->hasMany('PropertyDiscount', [
            'foreignKey' => 'discount_id',
            'joinTable' => 'property_discount'
        ]);

**Comment by markstory** [comment]
@cnizzardini I'll see if I can reproduce this and figure out where the hydrator is failing to build the required properties.

**Comment by cnizzardini** [comment]
On a hunch I swapped the key/value fields thinking it might want discount_id as the keyField, but no dice. 

Here is the SQL debug output from inside the callback if this helps at all:

SELECT `PropertyDiscount`.`property_id` AS `PropertyDiscount__property_id`, `PropertyDiscount`.`discount_id` AS `PropertyDiscount__discount_id` FROM `property_discount` `PropertyDiscount` WHERE `PropertyDiscount`.`discount_id` in (:c0,:c1,:c2,:c3,:c4,:c5,:c6,:c7,:c8,:c9,:c10,:c11)

**Comment by markstory** [comment]
> Is Cake having difficulty associating this back to its parent association when I specify a list finder?

Yes it is. The issue is that your `keyField` is a non-distinct field, this results in values being overwritten when the result set is built, causing the ORM not being able to attach the data to the source entity.

If you were to use an distinct field as the `keyField` you still  would have issues, as the ORM relies on the presence of the foreignKey fields in the association results in order to stitch records onto their parents. Because `find('list')` generates a simple key value store, this requirement cannot be met. If I'm understanding your question you'd like to get a list of `discount_id` values for each property. You can do this with a result formatter. Something like the following should give you what you want.

```php
 $query = $this->find()
	->contain(['PropertyDiscount'])
	->formatResults(function ($results) {
		return $results->map(function ($row) {
			if (count($row->property_discount)) {
				$row->property_discount = collection($row->property_discount)->extract('discount_id')->toList();
			}

			return $row;
		});
	});
```

**Comment by cnizzardini** [comment]
I assumed this was the problem. Yes my user-case requires that  a discount only be applied to properties defined as children of that discount. I had considered formatResults, but I am not sure how this is any more efficient that may current solution. I guess for code-reuse this is a better solution.

I've encountered issues before where if I do no select the foreign key field the ORM fails. Couldn't the ORM just "shadow select" the foreign key and alleviate developers from this problem?

**Comment by markstory** [comment]
> I've encountered issues before where if I do not select the foreign key field the ORM fails. Couldn't the ORM just "shadow select" the foreign key and alleviate developers from this problem?

This happens for some select based association types, but not join types. That improvement has an open issue #11580.

**Comment by cnizzardini** [comment]
@markstory I guess this should be closed then if #11580 will address this at a future date? Your call. At the very least this will show up in Google now and help other developers.

**Comment by markstory** [comment]
@cnizzardini ok. I think #11580 can be used to improve the queryBuilder to help warn or fix selected fields in situations like this.
